internal/services/containerapps/container_app_resource_test.go

```diff
@@ -548,6 +563,95 @@ resource "azurerm_container_app" "test" {
 `, r.template(data), data.RandomInteger)
 }
 
+func (r ContainerAppResource) withKeyVaultSecret(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {
+    key_vault {
+      purge_soft_delete_on_destroy    = true
+      recover_soft_deleted_key_vaults = true
+    }
+  }
+}
+
+%[1]s
+
+data "azurerm_client_config" "current" {}
+
+resource "azurerm_key_vault" "test" {
+  name                      = "acctest-kv-%[3]s"
+  resource_group_name       = azurerm_resource_group.test.name
+  location                  = azurerm_resource_group.test.location
+  tenant_id                 = data.azurerm_client_config.current.tenant_id
+  enable_rbac_authorization = true
+  sku_name                  = "standard"
+}
+
+resource "azurerm_key_vault_secret" "test" {
+  name         = "secret-%[3]s"
+  value        = "test-secret"
+  key_vault_id = azurerm_key_vault.test.id
+
+  depends_on = [
+    azurerm_role_assignment.self_key_vault_admin
+  ]
+}
+
+resource "azurerm_role_assignment" "self_key_vault_admin" {
+  scope                = azurerm_key_vault.test.id
+  role_definition_name = "Key Vault Administrator"
+  principal_id         = data.azurerm_client_config.current.object_id
+}
+
+resource "azurerm_role_assignment" "mi_key_vault_secrets" {
+  scope                = azurerm_key_vault.test.id
+  role_definition_name = "Key Vault Secrets User"
+  principal_id         = azurerm_user_assigned_identity.test.principal_id
+}
+
+resource "azurerm_user_assigned_identity" "test" {
+  name                = "acct-%[2]d"
+  resource_group_name = azurerm_resource_group.test.name
+  location            = azurerm_resource_group.test.location
+}
+
+resource "azurerm_container_app" "test" {
+  name                         = "acctest-capp-%[2]d"
+  resource_group_name          = azurerm_resource_group.test.name
+  container_app_environment_id = azurerm_container_app_environment.test.id
+  revision_mode                = "Single"
+
+  identity {
+    type         = "UserAssigned"
+    identity_ids = [azurerm_user_assigned_identity.test.id]
+  }
+
+  template {
+    container {
+      name   = "acctest-cont-%[2]d"
+      image  = "jackofallops/azure-containerapps-python-acctest:v0.0.1"
+      cpu    = 0.25
+      memory = "0.5Gi"
+      env {
+        name        = "key-vault-secret"
+        secret_name = "key-vault-secret"
+      }
+    }
+  }
+  secret {
+    name          = "key-vault-secret"
+    identity      = azurerm_user_assigned_identity.test.id
```

Comment:
Can we also test for the system assigned identity case?

Reply:
do we need this in a separate test or within the existing one?

Reply:
Can be in the same test case, but in some separate steps, e.g. UAI -> UAI,SAI -> SAI, so that we can test out the update logic.

Reply:
i've added `withKeyVaultSecretIdentityUpdate` which does:
- User Assigned
- User Assigned & System Assigned
- System Assigned

I've included the provisioning of the System Assigned identity within the User Assigned template (`withKeyVaultSecretUserIdentity`). This is as the API validates managed identity access to secrets when the container app is updated - so the identity and role assignment need to be created prior to the secret reference.

Reply:
Typically, this requires a separate TF resource which extract these key vault parts out, e.g. https://github.com/hashicorp/terraform-provider-azurerm/blob/7596b072d0b8bf04585d0b77facd8296640d8a2a/internal/services/kusto/kusto_cluster_customer_managed_key_resource.go. In the container app case, there probably wants a `azurerm_container_app_secret` resource, which allows users to manage a secret at one time.

With above, users then are able to provision the resources in the following order: container app -> role assignment to the container app system assigned identity -> one or more container app secrets

Reply:
okay cool i'll work on getting that created